path/path.go

```diff
@@ -3,188 +3,301 @@ package path
+	Root() cid.Cid
+
+	// Segments returns the different elements ofd a path, which are delimited
+	// by a forward slash ("/"). The leading slash must be ignored, that is, no
+	// segment should be empty.
+	Segments() []string
 }
 
-// FromCid safely converts a cid.Cid type to a Path type.
-func FromCid(c cid.Cid) Path {
-	return Path("/ipfs/" + c.String())
+// ResolvedPath is a [Path] which was resolved to the last resolvable node.
+type ResolvedPath interface {
+	Path
+
+	// Cid returns the [cid.Cid] of the node referenced by the path.
+	Cid() cid.Cid
+
+	// Remainder returns the unresolved parts of the path.
+	Remainder() string
 }
 
-// Segments returns the different elements of a path
-// (elements are delimited by a /).
-func (p Path) Segments() []string {
-	cleaned := path.Clean(string(p))
-	segments := strings.Split(cleaned, "/")
+// ImmutablePath is a [Path] which is guaranteed to return "false" to [Mutable].
+type ImmutablePath struct {
+	Path
+}
 
-	// Ignore leading slash
-	if len(segments[0]) == 0 {
-		segments = segments[1:]
+func NewImmutablePath(p Path) (ImmutablePath, error) {
+	if p.Mutable() {
+		return ImmutablePath{}, fmt.Errorf("path was expected to be immutable: %s", p.String())
 	}
 
-	return segments
+	return ImmutablePath{p}, nil
+}
+
+type path struct {
+	str       string
+	root      cid.Cid
+	namespace string
+}
+
+func (p *path) String() string {
+	return p.str
+}
+
+func (p *path) Namespace() string {
+	return p.namespace
 }
 
-// String converts a path to string.
-func (p Path) String() string {
-	return string(p)
+func (p *path) Mutable() bool {
+	return p.namespace == IPNSNamespace
 }
 
-// IsJustAKey returns true if the path is of the form <key> or /ipfs/<key>, or
-// /ipld/<key>
-func (p Path) IsJustAKey() bool {
-	parts := p.Segments()
-	return len(parts) == 2 && (parts[0] == "ipfs" || parts[0] == "ipld")
+func (p *path) Root() cid.Cid {
+	return p.root
 }
 
-// PopLastSegment returns a new Path without its final segment, and the final
-// segment, separately. If there is no more to pop (the path is just a key),
-// the original path is returned.
-func (p Path) PopLastSegment() (Path, string, error) {
+func (p *path) Segments() []string {
+	// Trim slashes from beginning and end, such that we do not return empty segments.
+	str := strings.TrimSuffix(p.str, "/")
+	str = strings.TrimPrefix(str, "/")
 
-	if p.IsJustAKey() {
-		return p, "", nil
+	return strings.Split(str, "/")
+}
+
+type resolvedPath struct {
+	path
+	cid       cid.Cid
+	remainder string
+}
+
+func (p *resolvedPath) Cid() cid.Cid {
+	return p.cid
+}
+
+func (p *resolvedPath) Remainder() string {
+	return p.remainder
+}
+
+// NewIPFSPath returns a new "/ipfs" path with the provided CID.
+func NewIPFSPath(cid cid.Cid) ResolvedPath {
+	return &resolvedPath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
+			root:      cid,
+			namespace: IPFSNamespace,
+		},
+		cid:       cid,
+		remainder: "",
 	}
+}
```

Comment:
```suggestion
type resolvedPath struct {
	path
}

func (p resolvedPath) Remainder() string {
	return ""
}

// NewIPFSPath returns a new "/ipfs" path with the provided CID.
func NewIPFSPath(cid cid.Cid) ResolvedPath {
	return resolvedPath{
		path: path{
			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
			root:      cid,
			namespace: IPFSNamespace,
		},
		cid:       cid,
		remainder: "",
	}
}
```
you are never returning something different from the underlying path, and the remainder is always empty so you can hardcode that